core/src/main/java/org/apache/druid/metadata/DynamicConfigProvider.java

```diff
@@ -0,0 +1,39 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.metadata;
+
+import com.fasterxml.jackson.annotation.JsonSubTypes;
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import org.apache.druid.guice.annotations.ExtensionPoint;
+
+import java.util.Map;
+
+/**
+ * This is used to get [secure] configuration in various places in an extensible way.
+ */
+@ExtensionPoint
+@JsonTypeInfo(use = JsonTypeInfo.Id.NAME, property = "type", defaultImpl = MapStringDynamicConfigProvider.class)
+@JsonSubTypes(value = {
+    @JsonSubTypes.Type(name = "mapString", value = MapStringDynamicConfigProvider.class),
+})
+public interface DynamicConfigProvider<T>
```

Comment:
nit: should we annotate `PasswordProvider` as `@Deprecated` in favor of this class?

Reply:
I didn't mark that deprecated because users wanting to have their own extension for db password must still implement a `PasswordProvider` ,
however, for Druid devs, any new credential/extensible-config type thing must use `DynamicConfigProvider` ... so `PasswordProvider` is "deprecated" in that sense.
will mark it deprecated and add few comments.